torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -314,28 +314,9 @@ class ProcessGroupNCCL : public ProcessGroup {
 
       // Do not free the underlying data storage of value_ before its
       // usage on futureNCCLCallbackStream_ finish.
-      if (record_stream_cb_ != nullptr) {
-        // If a Python communication hook is used, record_stream_cb_ will be
-        // set in torch/csrc/jit/python/pybind_utils.h, which allows Python
-        // dependency to be imported.
-        record_stream_cb_(value_, futureNCCLCallbackStream_->unwrap());
-      } else {
-        // If a C++ communication hook is used, create and set a record stream
-        // callback.
-        TORCH_INTERNAL_ASSERT(
-            value_.isTensorList() || value_.isTensor(),
-            "the future value must be either a tensor list or a tensor.");
-        at::Tensor tensor;
-        if (value_.isTensorList()) {
-          const auto tensors = value_.toTensorVector();
-          TORCH_INTERNAL_ASSERT(
-              tensors.size() == 1, "expected exactly 1 tensor");
-          tensor = tensors[0];
-        } else {
-          tensor = value_.toTensor();
-        }
+      for (const at::DataPtr& data_ptr : extractDataPtrs(value_)) {
         c10::cuda::CUDACachingAllocator::recordStream(
```

Comment:
Curious about the implication for RPC use cases. Do RPC also need to call `recordStream`? If yes, when? Is it when the tensors are retrieved from the Future (through `result` or `wait`), we should call `recordStream` on the current stream?

Reply:
I'm still figuring out the correct usage of the caching allocator, but I think this should work in the same way for RPC. The model I have in mind for RPC is the one we discussed in https://github.com/pytorch/pytorch/issues/44084#issuecomment-688313258. In that case, in the receiver (bottom right quadrant of the diagram), I think we need to record streams with the caching allocators whenever we "transfer" the result to other streams than the ones we used to receive it. This would happen both when using `.wait()`/`.value()`, and in callbacks (basically the points in the diagram where we say "record events"). And this is exactly what we're doing here. Does this make sense?

Reply:
yep, make sense to me.